esp/logging/logginglib/logthread.cpp

```diff
@@ -646,39 +646,62 @@ void CLogRequestReader::removeUnknownAckedLogFiles(StringArray& ackedLogFiles)
     ESPLOG(LogMax, "#### The removeUnknownAckedLogFiles() done.");
 }
 
+void CLogRequestReader::addNewAckedFileList(const char* list, StringArray& fileNames)
+{
+    OwnedIFile newList = createIFile(list);
+    if (!newList)
+        throw makeStringExceptionV(EspLoggingErrors::UpdateLogFailed, "Failed to access %s", list);
+    if (newList->exists())
+    {
+        newList->remove();
+        newList.setown(createIFile(list));
+    }
+    OwnedIFileIO newListIO = newList->open(IFOwrite);
+    if (!newListIO)
+        return; //Should never happen
+
+    offset_t pos = 0;
+    ForEachItemIn(i, fileNames)
+    {
+        const char* fileName = fileNames.item(i);
+        StringBuffer line(fileName);
+        line.append("\r\n");
+
+        unsigned len = line.length();
+        newListIO->write(pos, len, line.str());
+        pos += len;
+        PROGLOG("Add AckedLogFile %s to %s", fileName, list);
+    }
+}
 //The file names in the fileNames should be removed from both ackedLogFileCheckList
 //and settings->ackedFileList.
 void CLogRequestReader::cleanAckedLogFiles(StringArray* fileNames)
 {
     CriticalBlock b(crit);
 
-    OwnedIFile ackedFiles = createIFile(settings->ackedFileList);
-    if (!ackedFiles)
-        throw makeStringExceptionV(EspLoggingErrors::UpdateLogFailed, "Failed to access %s", settings->ackedFileList.str());
-
-    //Clean the disk file: settings->ackedFileList.
-    ackedFiles->remove();
-    PROGLOG("Clean %s", settings->ackedFileList.str());
+    //Find which file should not be removed from ackedLogFileCheckList.
+    StringArray fileNamesToKeep;
+    for (auto r : ackedLogFileCheckList)
+    {
+        if (!fileNames->contains(r.c_str()))
+            fileNamesToKeep.append(r.c_str());
+    }
 
-    //Remove the file names from ackedLogFileCheckList
-    ForEachItemIn(i, *fileNames)
-        ackedLogFileCheckList.erase(fileNames->item(i));
+    //Create a temp file with the fileNamesToKeep for replacing the settings->ackedFileList
+    VStringBuffer tempFileName("%s.tmp", settings->ackedFileList.str());
+    addNewAckedFileList(tempFileName, fileNamesToKeep);
 
-    //Synchronize the new ackedLogFileCheckList to the settings->ackedFileList.
-    OwnedIFileIO ackedFilesIO = ackedFiles->open(IFOwrite);
-    if (!ackedFilesIO)
-        return; //Should never happend
+    //Replace the settings->ackedFileList with the temp file
+    renameFile(settings->ackedFileList, tempFileName, true);
```

Comment:
I think the rename needs to be conditional on the currently non-existent result of addNewAckedFileList to be considered safe. Otherwise, the temp file is just additional overhead.